### PR TITLE
Add dotnet immediate accept adapter sample to experimental

### PR DESCRIPTION
**Pull request by EricDahlvang**
in process guidance for https://github.com/microsoft/botframework-sdk/issues/5500 

addresses: https://github.com/microsoft/botframework-sdk/issues/5665

### CODE REVIEW

File: experimental/immediate-accept-adapter/csharp_dotnetcore/ImmediateAcceptAdapter.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        /// Note, this is an ImmediateAccept and BackgroundProcessing replacement for: 
+        /// Task ProcessAsync(HttpRequest httpRequest, HttpResponse httpResponse, IBot bot, CancellationToken cancellationToken = default);
+        /// </summary>
+        /// <param name="httpRequest">The HTTP request object, typically in a POST handler by a Controller.</param>
+        /// <param name="httpResponse">The HTTP response object.</param>
+        /// <param name="bot">The bot implementation.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive
+        ///     notice of cancellation.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        public async Task ProcessOnBackgroundThreadAsync(HttpRequest httpRequest, HttpResponse httpResponse, IBot bot, CancellationToken cancellationToken = default)
```
Comment (johnataylor) [inline experimental/immediate-accept-adapter/csharp_dotnetcore/ImmediateAcceptAdapter.cs:63]: Do you need a different method here?
I mean you could just override and delegate calls to the base class appropriately.
(In other words the relationship with the Controller - which is what the existing interface is modeling - technically hasn't changed - so I don't need a new type exposed to the Controller. OK I realize the semantics have changed slightly. But the semantics of the protocol are a little unconventional - and not directly modeled - anyhow.)

File: experimental/immediate-accept-adapter/csharp_dotnetcore/BackgroundActivityService/IActivityTaskQueue.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+{
+    /// <summary>
+    /// Interface for a class used to transfer an AcctivityWithClaims to the <see cref="HostedActivityService"/>.
+    /// </summary>
+    public interface IActivityTaskQueue
+    {
+        /// <summary>
+        /// Enqueue an Activity, with Claims, to be processed on a background thread.
+        /// 
+        /// NOTE: It is assumed these claims have been authenticated via JwtTokenValidation.AuthenticateRequest
```
Comment (EricDahlvang) [inline experimental/immediate-accept-adapter/csharp_dotnetcore/BackgroundActivityService/IActivityTaskQueue.cs:19]: change these to remarks

File: experimental/immediate-accept-adapter/csharp_dotnetcore/BackgroundActivityService/HostedActivityService.cs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+                {
+                    try
+                    {
+                        // The read lock will not be acquirable if the app is shutting down.
+                        // New tasks should not be starting during shutdown.
+                        if (_lock.TryEnterReadLock(500))
+                        {
+                            // Create the task which will execute the work item.
+                            var task = GetTaskFromWorkItem(activityWithClaims, stoppingToken)
+                                .ContinueWith(t =>
```
Comment (mrivera-ms) [inline experimental/immediate-accept-adapter/csharp_dotnetcore/BackgroundActivityService/HostedActivityService.cs:112]: Pass the cancellation token.

File: experimental/immediate-accept-adapter/csharp_dotnetcore/BackgroundTaskService/HostedTaskService.cs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+namespace ImmediateAcceptBot.BackgroundQueue
+{
+    /// <summary>
+    /// BackgroundService implementation used to process work items on background threads.
+    /// 
+    /// <seealso cref="https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.backgroundservice"/>
```
Comment (mrivera-ms) [inline experimental/immediate-accept-adapter/csharp_dotnetcore/BackgroundTaskService/HostedTaskService.cs:18]: Use href.

File: experimental/immediate-accept-adapter/csharp_dotnetcore/BackgroundTaskService/HostedTaskService.cs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        private readonly ConcurrentDictionary<Func<CancellationToken,Task>, Task> _tasks = new ConcurrentDictionary<Func<CancellationToken, Task>, Task>();
+        private readonly IBackgroundTaskQueue _taskQueue;
+        private readonly int _shutdownTimeoutSeconds;
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="config"></param>
+        /// <param name="taskQueue"></param>
+        /// <param name="logger"></param>
```
Comment (mrivera-ms) [inline experimental/immediate-accept-adapter/csharp_dotnetcore/BackgroundTaskService/HostedTaskService.cs:33]: Please add missing docs.

File: experimental/immediate-accept-adapter/csharp_dotnetcore/Bots/EchoBot.cs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+            _shutdownTimeoutSeconds = config.GetValue<int>("ShutdownTimeoutSeconds");
+            _botId = config.GetValue<string>("MicrosoftAppId");
+            _botId = string.IsNullOrEmpty(_botId) ? Guid.NewGuid().ToString() : _botId;
+            _taskQueue = taskQueue;
+        }
+
+        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
+        {
+            var text = turnContext.Activity.Text?.ToLower()?.Trim();
```
Comment (mrivera-ms) [inline experimental/immediate-accept-adapter/csharp_dotnetcore/Bots/EchoBot.cs:44]: Don't need null check in ToLower().

File: experimental/immediate-accept-adapter/csharp_dotnetcore/Bots/EchoBot.cs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+                    await turnContext.SendActivityAsync($"okay, I will background message you after: {seconds} seconds", cancellationToken: cancellationToken).ConfigureAwait(false);
+                    _taskQueue.QueueBackgroundWorkItem(async cancelToken => await ProactiveMessageCallbackAsync(turnContext.Adapter, turnContext.Activity.GetConversationReference(), message, seconds, cancelToken).ConfigureAwait(false));
+
+                }
+            }
+        }
+
+        private async Task ProactiveMessageCallbackAsync(BotAdapter adapter, ConversationReference conversationReference, string message, int seconds, CancellationToken cancellationToken)
+        {
+            // Pause on the background thread for the number of seconds specified, then load the conversationi and message the user.
```
Comment (mrivera-ms) [inline experimental/immediate-accept-adapter/csharp_dotnetcore/Bots/EchoBot.cs:110]: Typo: conversation.

File: experimental/immediate-accept-adapter/csharp_dotnetcore/ImmediateAcceptAdapter.cs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+using Microsoft.Bot.Connector.Authentication;
+using Microsoft.Bot.Schema;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+
+namespace ImmediateAcceptBot
+{
+    public class ImmediateAcceptAdapter : BotFrameworkHttpAdapter
+    {
+        private readonly IBackgroundTaskQueue _backgroundTaskQueue;
```
Comment (mrivera-ms) [inline experimental/immediate-accept-adapter/csharp_dotnetcore/ImmediateAcceptAdapter.cs:24]: This is never used.

File: experimental/immediate-accept-adapter/csharp_dotnetcore/ImmediateAcceptAdapter.cs [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+                await turnContext.TraceActivityAsync("OnTurnError Trace", exception.Message, "https://www.botframework.com/schemas/error", "TurnError");
+            };
+        }
+
+        /// <summary>
+        /// This method can be called from inside a POST method on any Controller implementation.  If the activity is Not an Invoke, and
+        /// DeliveryMode is Not ExpectReplies, and this is not a Get request to upgrade to WebSockets, then the activity will be enqueued
+        /// to be processed on a background thread.
+        /// 
+        ///
```
Comment (mrivera-ms) [inline experimental/immediate-accept-adapter/csharp_dotnetcore/ImmediateAcceptAdapter.cs:52]: Remove extra lines.

File: experimental/immediate-accept-adapter/csharp_dotnetcore/ImmediateAcceptAdapter.cs (hunk 1 above)
Comment (EricDahlvang) [reply experimental/immediate-accept-adapter/csharp_dotnetcore/ImmediateAcceptAdapter.cs:63]: @johnataylor  Should we make ProcessAsync virtual or protected?

Currently:
```cs
public Task ProcessAsync(HttpRequest httpRequest, HttpResponse httpResponse, IBot bot, CancellationToken cancellationToken = default);
```

**Pull request merged**